### ISSUE TITLE
[Feature] Improve time-dependent Hamiltonian interface for arrays of durations

### ISSUE THREAD

**Issue by jpmoutinho** [pasqal-io/qadence#601]
### Describe the feature

Currently when defining a time-dependent `HamEvo` the `duration` must be hard-coded as a float when instantiating the operator. Instead, it should follow the typical parameter interface for Qadence, including the option to pass it as a `FeatureParameter` (or even `VariationalParameter`) so that it can be defined at execution instead. 

A few ideas:
- To be discussed, but I think defining the `duration` as a `FeatureParameter` should be the default behaviour when no `duration` is passed. Then, when doing `evolution = HamEvo(time_generator, t)` and running the `evolution`, Qadence will complain that there is a missing `"duration"` in the values dict.
- As with other parameters in Qadence, batching should be supported in values dict entry for `duration`. However, the batch of duration values should be automatically sorted, and then the solver should simply output the desired value when the timestep reaches each value in the batch. I understand there is a problem here with matching the exact value in the input batch with the actual value the solver uses (because I guess it depends on the timesteps). Not sure what is the best solution for this. Maybe the solver should just output the value for the timestep that is closest to that duration value. Anyway, the point would be that doing `run(evolution, values = {"duration": [100.0]}` or `run(evolution, values = {"duration": [0.0, 20., 40., 60., 80., 100.]}` should take roughly the same.

Maybe the second point can be done in a subsequent MR to the first.

### Would you like to work on this issue?

None

**Comment by jpmoutinho** [comment]
@RolandMacDoland this would be an important change to have for the time-dependent hamiltonians. Currently it's highly inefficient to measure something like an expectation value over time, because there's no way to access the results in the middle time steps.

**Comment by vytautas-a** [comment]
Agreed. It should not be difficult to achieve since the underlying time-dependent solvers already return the state values for each time moment and the corresponding times.

**Comment by RolandMacDoland** [comment]
@vytautas-a Are you OK to pick that one up in the next sprint ?
